### PR TITLE
TLS tunnel over Lwt_io.channel

### PR DESCRIPTION
**Pull request by art-w**
I'm marking this PR as a draft, because it exposes the functionalities of TLS-over-TLS added by this PR https://github.com/mirleft/ocaml-tls/pull/499 which hasn't yet been reviewed... but I have some questions on how to best present this feature in Conduit :)

For context, @MisterDA has a working implementation of `cohttp-lwt-unix` client requests with http/https proxying, which relies on `conduit-lwt-unix` for TLS and on this PR for the "https over http(s)", which happens because https proxies provide a safe tunnel for the client to negotiate the TLS connection with the server. For this to work though, we need to able to reuse an existing connection instead of having Conduit create a new socket. The https://github.com/mirleft/ocaml-tls/pull/499 PR has more details on how this works, but I've also added an example in Conduit to test the flow if you are curious.

### CODE REVIEW

File: src/conduit-lwt-unix/conduit_lwt_unix.ml [hunk 1]
```diff
@@ -410,6 +423,7 @@ let serve ?backlog ?timeout ?stop ~on_exn ~(ctx : ctx) ~(mode : server) callback
 
 let endp_of_flow = function
   | TCP { ip; port; _ } -> `TCP (ip, port)
+  | Tunnel -> `Unknown "TLS tunnel"
```
Comment (art-w) [inline src/conduit-lwt-unix/conduit_lwt_unix.ml:426]: I'm not sure what is the best way to address this? We can't expose `Lwt_io.channel` in `Conduit.endp` obviously, should `conduit-lwt-unix` expose an extended `endp` type with the new tunnel option?

**Comment by hannesm** [comment]
What state is this in? Is there anyone interested in conduit reviewing feature PRs? I still don't understand conduits purpose, and only do maintenance (mainly because I want to release mirage) ;)

**Comment by art-w** [comment]
I've rebased and removed the draft status, since the tls dependency has since been released :)
The feature has been thoroughly tested with https://github.com/mirage/ocaml-cohttp/pull/1080 so I think it's in good shape, although it would be nice to have a conduit expert opinion on the `endp_of_flow` question.

File: src/conduit-lwt-unix/conduit_lwt_unix.ml (hunk 1 above)
Comment (art-w) [reply src/conduit-lwt-unix/conduit_lwt_unix.ml]: I went ahead and exposed a new `Conduit_lwt_unix.endp` which extends `Conduit.endp` with the Lwt-specific TLS tunneling

**Comment by art-w** [comment]
I went with exposing the TLS tunneling over an existing `Lwt_io.channel` connection by exposing a new type `Conduit_lwt_unix.endp` which extends `Conduit.endp` with the Lwt-specific feature, as this seemed like the most natural solution... @samoht Would you have some time to look at this PR? :)

File: src/conduit-lwt-unix/conduit_lwt_unix.ml [hunk 2]
```diff
@@ -262,31 +264,41 @@ let set_max_active maxactive = Conduit_lwt_server.set_max_active maxactive
...
+      >|= fun certificate -> Some (`Single certificate)
+
+let domain_name hostname =
+  try Domain_name.(host_exn (of_string_exn hostname))
+  with Invalid_argument msg ->
+    let s =
+      Printf.sprintf "couldn't convert %s to a [`host] Domain_name.t: %s"
+        hostname msg
+    in
+    invalid_arg s
```
Comment (MisterDA) [inline src/conduit-lwt-unix/conduit_lwt_unix.ml:283]: You can preserve the backtrace of the exception if you intent to re-raise (even with a modified message) it:

```suggestion
    let trace = Printexc.get_raw_backtrace () in
    let msg =
      Printf.sprintf "couldn't convert %s to a [`host] Domain_name.t: %s"
        hostname msg
    in
    Printexc.raise_with_backtrace (Invalid_argument msg) trace
```

**Review by MisterDA — APPROVED** [review]

**Comment by samoht** [comment]
LGTM

**Pull request merged**